File: backend/location-advisory/streaming/consumer.py
import json
import logging
import redis
from confluent_kafka import Consumer, KafkaError

logger = logging.getLogger(__name__)

class SwitchEventConsumer:
    """
    Kafka consumer that listens to NIE switch events on 'network.switch.events'.
    It extracts the historical signal strength logged during the switch
    and aggregates it into the appropriate H3 grid cell in Redis.
    """
    
    def __init__(self, kafka_config: dict, redis_client: redis.Redis):
        self.consumer = Consumer(kafka_config)
        self.redis = redis_client
        self.topic = "network.switch.events"
        
    def start(self):
        self.consumer.subscribe([self.topic])
        logger.info(f"Subscribed to {self.topic}")
        
        try:
            while True:
                msg = self.consumer.poll(1.0)
                
                if msg is None:
                    continue
                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        continue
                    else:
                        logger.error(f"Kafka error: {msg.error()}")
                        break

                self._process_message(msg.value())
                
        except KeyboardInterrupt:
            logger.info("Kafka consumer shutting down...")
        finally:
            self.consumer.close()

    def _process_message(self, message_bytes: bytes):
        try:
            data = json.loads(message_bytes.decode('utf-8'))
            
            # Extract H3 location if the NIE attached it, or calculate if GPS included
            # In LLD, NIE optionally provides grid data directly via client SDK payload
            h3_index = data.get("h3_index")
            if not h3_index:
                return # Can't aggregate without location
                
            scores = json.loads(data.get("quality_scores_json", "{}"))
            
            # Update EMA (Exponential Moving Average) in Redis for this hex
            for carrier, score in scores.items():
                self._update_ema(h3_index, carrier, float(score))
                
        except Exception as e:
            logger.error(f"Error processing message: {e}")

    def _update_ema(self, h3_index: str, carrier: str, new_score: float):
        """
        Calculates an Exponential Moving Average for the grid cell to keep the
        advisory score smooth and resistant to single-device anomalies.
        """
        key = f"advisory:h3:{h3_index}:{carrier}"
        alpha = 0.2
        
        current = self.redis.get(key)
        if current is None:
            self.redis.set(key, new_score, ex=86400 * 7) # 7 day TTL
        else:
            ema = (new_score * alpha) + (float(current) * (1 - alpha))
            self.redis.set(key, ema, ex=86400 * 7)

File: backend/location-advisory/consumer/consumer.go
package consumer

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
	"github.com/redis/go-redis/v9"
)

type SwitchEventConsumer struct {
	consumer    *kafka.Consumer
	redisClient *redis.Client
	topic       string
}

func NewSwitchEventConsumer(bootstrapServers, groupID string, redisClient *redis.Client) (*SwitchEventConsumer, error) {
	c, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": bootstrapServers,
		"group.id":          groupID,
		"auto.offset.reset": "earliest",
	})
	if err != nil {
		return nil, err
	}

	return &SwitchEventConsumer{
		consumer:    c,
		redisClient: redisClient,
		topic:       "network.switch.events",
	}, nil
}

func (c *SwitchEventConsumer) Start(ctx context.Context) {
	err := c.consumer.SubscribeTopics([]string{c.topic}, nil)
	if err != nil {
		log.Printf("Failed to subscribe to topic: %v", err)
		return
	}

	log.Printf("Subscribed to %s", c.topic)

	for {
		select {
		case <-ctx.Done():
			log.Println("Kafka consumer shutting down...")
			c.consumer.Close()
			return
		default:
			msg, err := c.consumer.ReadMessage(1 * time.Second)
			if err == nil {
				c.processMessage(msg.Value)
			} else if !err.(kafka.Error).IsTimeout() {
				log.Printf("Kafka error: %v (%v)", err, msg)
			}
		}
	}
}

func (c *SwitchEventConsumer) processMessage(messageBytes []byte) {
	var data map[string]interface{}
	if err := json.Unmarshal(messageBytes, &data); err != nil {
		log.Printf("Error processing message: %v", err)
		return
	}

	h3Index, ok := data["h3_index"].(string)
	if !ok || h3Index == "" {
		return // Can't aggregate without location
	}

	qualityScoresStr, ok := data["quality_scores_json"].(string)
	if !ok {
		return
	}

	var scores map[string]float64
	if err := json.Unmarshal([]byte(qualityScoresStr), &scores); err != nil {
		log.Printf("Error processing quality scores: %v", err)
		return
	}

	for carrier, score := range scores {
		c.updateEMA(h3Index, carrier, score)
	}
}

func (c *SwitchEventConsumer) updateEMA(h3Index string, carrier string, newScore float64) {
	key := fmt.Sprintf("advisory:h3:%s:%s", h3Index, carrier)
	alpha := 0.2

	ctx := context.Background()
	currentStr, err := c.redisClient.Get(ctx, key).Result()
	if err == redis.Nil {
		c.redisClient.Set(ctx, key, newScore, 7*24*time.Hour)
	} else if err == nil {
		var current float64
		fmt.Sscanf(currentStr, "%f", &current)
		ema := (newScore * alpha) + (current * (1 - alpha))
		c.redisClient.Set(ctx, key, ema, 7*24*time.Hour)
	} else {
		log.Printf("Error getting redis key %s: %v", key, err)
	}
}
